File: api/predictions/predict.py
"""POST /api/predictions/predict"""
import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

import json
import uuid
from datetime import datetime
from http.server import BaseHTTPRequestHandler

from _shared.auth import auth_enabled, get_user_id
from _shared.db import save_prediction
from _shared.predictor import predict_protein_structure
from _shared.response import json_response, options_response

MAX_SEQUENCE_LENGTH = int(os.environ.get('MAX_SEQUENCE_LENGTH', 10000))
VALID_AA = set('ACDEFGHIKLMNPQRSTVWY')


class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        options_response(self)

    def do_POST(self):
        # Auth check
        user_id = get_user_id(dict(self.headers))
        if auth_enabled() and not user_id:
            return json_response(self, 401, {
                'status': 'error',
                'message': 'Authentication required. Sign in with Supabase.',
            })

        # Parse body
        length = int(self.headers.get('Content-Length', 0))
        body = {}
        if length:
            try:
                body = json.loads(self.rfile.read(length))
            except Exception:
                return json_response(self, 400, {'status': 'error', 'message': 'Invalid JSON body'})

        sequence = body.get('sequence', '')

        # Validation
        if not sequence:
            return json_response(self, 400, {'status': 'error', 'message': 'Sequence is required'})
        if len(sequence) < 2:
            return json_response(self, 400, {
                'status': 'error',
                'message': 'Sequence too short. Minimum length is 2 amino acids.',
            })
        if len(sequence) > MAX_SEQUENCE_LENGTH:
            return json_response(self, 400, {
                'status': 'error',
                'message': f'Sequence too long ({len(sequence)} residues). Maximum allowed is {MAX_SEQUENCE_LENGTH}.',
            })
        sequence = sequence.upper()
        if not all(c in VALID_AA for c in sequence):
            return json_response(self, 400, {
                'status': 'error', 'message': 'Invalid amino acids in sequence',
            })

        # Predict
        prediction_id = f'pred_{uuid.uuid4().hex[:12]}'
        result = predict_protein_structure(sequence)
        timestamp = datetime.now().isoformat()

        # Save
        if not save_prediction(
            pred_id=prediction_id,
            sequence=sequence,
            structure_info=result['structure'],
            coordinates=result['coordinates'],
            binding_pockets=result['binding_pockets'],
            model_source=result['model_source'],
            timestamp=timestamp,
            user_id=user_id,
        ):
            return json_response(self, 500, {
                'status': 'error', 'message': 'Failed to save prediction to database.',
            })

        json_response(self, 200, {
            'prediction_id': prediction_id,
            'sequence': sequence,
            'length': len(sequence),
            'structure': result['structure'],
            'coordinates': result['coordinates'],
            'binding_pockets': result['binding_pockets'],
            'model_source': result['model_source'],
            'status': 'success',
            'timestamp': timestamp,
        })

    def log_message(self, *args):
        pass

File: api/predictions/predict.js
import { predictProteinStructure } from '../utils/coordinate_generator.js';
import { supabase, getUserFromAuth } from '../utils/supabase.js';

export default async function handler(req, res) {
  if (req.method !== 'POST') return res.status(405).json({ message: 'Method Not Allowed' });

  try {
    const { sequence } = req.body || {};
    if (!sequence) return res.status(400).json({ status: 'error', message: 'Sequence is required' });

    const maxLen = 10000;
    const minLen = 2;

    if (sequence.length < minLen) {
      return res.status(400).json({ status: 'error', message: `Sequence too short. Min is ${minLen}` });
    }
    if (sequence.length > maxLen) {
      return res.status(400).json({ status: 'error', message: `Sequence too long. Max is ${maxLen}` });
    }

    if (!/^[ACDEFGHIKLMNPQRSTVWY]+$/i.test(sequence)) {
      return res.status(400).json({ status: 'error', message: 'Invalid amino acids' });
    }

    const cleanSeq = sequence.toUpperCase();
    const user = await getUserFromAuth(req);

    if (!user) {
      return res.status(401).json({ status: 'error', message: 'Authenticated user required' });
    }

    // Check cache
    const { data: cached } = await supabase
      .from('predictions')
      .select('*')
      .eq('sequence', cleanSeq)
      .limit(1)
      .single();

    let result;
    if (cached && cached.structure_info) {
      result = {
        structure: cached.structure_info,
        coordinates: cached.coordinates,
        binding_pockets: cached.binding_pockets,
        model_source: cached.model_source || 'Cache',
      };
    } else {
      result = await predictProteinStructure(cleanSeq);
    }

    const predictionId = `pred_${Math.random().toString(36).substring(2, 14)}`;
    const timestamp = new Date().toISOString();

    const { error } = await supabase.from('predictions').insert({
      id: predictionId,
      sequence: cleanSeq,
      length: cleanSeq.length,
      structure_info: result.structure,
      coordinates: result.coordinates,
      binding_pockets: result.binding_pockets,
      model_source: result.model_source,
      user_id: user.id,
      created_at: timestamp,
      updated_at: timestamp,
    });

    if (error) {
      console.error('Supabase Insert Error:', error);
      return res.status(500).json({ status: 'error', message: `Failed to save prediction to DB: ${error.message}` });
    }

    return res.status(200).json({
      prediction_id: predictionId,
      sequence: cleanSeq,
      length: cleanSeq.length,
      ...result,
      status: 'success',
      timestamp,
    });
  } catch (error) {
    console.error(error);
    return res.status(500).json({ status: 'error', message: 'Internal server error' });
  }
}
